test/e2e/autoscaling/horizontal_pod_autoscaling.go

```diff
@@ -168,3 +180,88 @@ func scaleDown(name string, kind schema.GroupVersionKind, checkStability bool, f
 	}
 	scaleTest.run(name, kind, f)
 }
+
+type HPAContainerResourceScaleTest struct {
+	initPods                    int
+	totalInitialCPUUsage        int
+	perPodCPURequest            int64
+	targetCPUUtilizationPercent int32
+	minPods                     int32
+	maxPods                     int32
+	noScale                     bool
+	noScaleStasis               time.Duration
+	firstScale                  int
+	firstScaleStasis            time.Duration
+	cpuBurst                    int
+	secondScale                 int32
+	runSidecar                  bool
+	busySidecar                 bool
+}
+
+func (scaleTest *HPAContainerResourceScaleTest) run(name string, kind schema.GroupVersionKind, f *framework.Framework) {
+	const timeToWait = 15 * time.Minute
+	rc := e2eautoscaling.NewDynamicResourceConsumer(name, f.Namespace.Name, kind, scaleTest.initPods, scaleTest.totalInitialCPUUsage, 0, 0, scaleTest.perPodCPURequest, 200, f.ClientSet, f.ScalesGetter, scaleTest.runSidecar, scaleTest.busySidecar)
+	defer rc.CleanUp()
+	hpa := e2eautoscaling.CreateContainerResourceCPUHorizontalPodAutoscaler(rc, scaleTest.targetCPUUtilizationPercent, scaleTest.minPods, scaleTest.maxPods)
+	defer e2eautoscaling.DeleteContainerResourceHPA(rc, hpa.Name)
+
+	if scaleTest.noScale {
+		if scaleTest.noScaleStasis > 0 {
+			rc.EnsureDesiredReplicasInRange(scaleTest.initPods, scaleTest.initPods, scaleTest.noScaleStasis, hpa.Name)
+		}
+	} else {
+		rc.WaitForReplicas(scaleTest.firstScale, timeToWait)
+		if scaleTest.firstScaleStasis > 0 {
+			rc.EnsureDesiredReplicasInRange(scaleTest.firstScale, scaleTest.firstScale+1, scaleTest.firstScaleStasis, hpa.Name)
+		}
+		if scaleTest.cpuBurst > 0 && scaleTest.secondScale > 0 {
+			rc.ConsumeCPU(scaleTest.cpuBurst)
+			rc.WaitForReplicas(int(scaleTest.secondScale), timeToWait)
+		}
+	}
+}
+
+func scaleOnIdleSideCar(name string, kind schema.GroupVersionKind, checkStability bool, f *framework.Framework) {
+	// Scale up on a busy application with an idle sidecar container
+	stasis := 0 * time.Minute
+	if checkStability {
+		stasis = 10 * time.Minute
+	}
+	scaleTest := &HPAContainerResourceScaleTest{
+		initPods:                    1,
+		totalInitialCPUUsage:        250,
+		perPodCPURequest:            500,
+		targetCPUUtilizationPercent: 20,
+		minPods:                     1,
+		maxPods:                     5,
+		firstScale:                  3,
+		firstScaleStasis:            stasis,
+		cpuBurst:                    700,
+		secondScale:                 5,
```

Comment:
Can you show me the math for why 5 is the correct result?  I would expect half of the `perPodCPURequest` to go to the sidecar (even split) and the load to be placed entirely on the non-sidecar container.  So somehow the load + the 250m CPU request should ==> 5 replicas.  Right?

Reply:
This is because I am not diving the CPU request in half and I am using the same value for the sidecar container as entire pod 
```
f sidecarStatus == Enable {
		additionalContainers = []v1.Container{
			{
				Name:    name + "-sidecar",
				Image:   resourceConsumerImage,
				Command: []string{"/consumer", "-port=8081"},
				Ports:   []v1.ContainerPort{{ContainerPort: 80}},
			},
		}

		if cpuLimit > 0 || memLimit > 0 {
			additionalContainers[0].Resources.Limits = v1.ResourceList{}
			additionalContainers[0].Resources.Requests = v1.ResourceList{}
		}

		if cpuLimit > 0 {
			additionalContainers[0].Resources.Limits[v1.ResourceCPU] = *resource.NewMilliQuantity(cpuLimit, resource.DecimalSI)
			additionalContainers[0].Resources.Requests[v1.ResourceCPU] = *resource.NewMilliQuantity(cpuLimit, resource.DecimalSI)
		}

		if memLimit > 0 {
			additionalContainers[0].Resources.Limits[v1.ResourceMemory] = *resource.NewQuantity(memLimit*1024*1024, resource.DecimalSI)
			additionalContainers[0].Resources.Requests[v1.ResourceMemory] = *resource.NewQuantity(memLimit*1024*1024, resource.DecimalSI)
		}
```
should I split them in half ?

Reply:
If I understand this correct, the parameter `scaleTest.perPodCPURequest` is used for both the resource consumer and the busy or idle sidecar.  So it's actually `perContainerCPURequest`.

Reply:
yes

Reply:
I will update the field name in the struct.

Reply:
Updated